### ISSUE TITLE
ANON_AVG has a strong bias in some situation

### ISSUE THREAD

**Issue by sebastian-sohr** [google/differential-privacy#6]
Hallo,
I tested the DP aggregation functions of the postgres plugin on a table with uniformly distributed numbers of type 0.00, 0.01, ..., 9.99, 10.00.

In the following situation I get results for ANON_AVG which are obviously not correct:

```
dp_test=> select
        avg(a),
        anon_avg(a),
        anon_avg_with_bounds(a, 0.0, 10.0),
        anon_sum(a) / anon_count(a) as anan_avg_manual,
        count(a),
        anon_count(a),
        sum(a),
        anon_sum(a),
        anon_ntile(a, 0.5, 0.0, 10.0)
from
        A
where
        b between 3.00 and 3.01
;
        avg         |     anon_avg     | anon_avg_with_bounds | anan_avg_manual  | count | anon_count |   sum    | anon_sum |    anon_ntile    
--------------------+------------------+----------------------+------------------+-------+------------+----------+----------+------------------
 5.1141938674579624 | 4.66269841269841 |     5.11089108910891 | 5.11177052423343 |  2022 |       2022 | 10340.90 |    10336 | 5.13447173044956
(1 row)

```
Obviously, the number 4.66269841269841 for the anon_avg is much to small, compared with the non anonymized values as well as the values of ana_avg_with_bounds and anon_sum() / anon_sum().
I had this strong bias in many repeats.

Strangly, for bigger sample, this "outlier" does not appear:

```
dp_test=> select
        avg(a),
        anon_avg(a),
        anon_avg_with_bounds(a, 0.0, 10.0),
        anon_sum(a) / anon_count(a) as anan_avg_manual,
        count(a),
        anon_count(a),
        sum(a),
        anon_sum(a),
        anon_ntile(a, 0.5, 0.0, 10.0)
from
        A
where
        b between 3.00 and 3.10
;

        avg         |     anon_avg     | anon_avg_with_bounds | anan_avg_manual  | count | anon_count |   sum    | anon_sum |    anon_ntile    
--------------------+------------------+----------------------+------------------+-------+------------+----------+----------+------------------
 5.0534345859108116 | 5.05230652013299 |     5.05318559556787 | 5.05014313417675 | 10831 |      10829 | 54733.75 |    54688 | 5.08097481360734
(1 row)
```

Strange.

Regards,
Sebastian.

**Comment by sebastian-sohr** [comment]
Oh, the restriction `b between 3.00 and 3.10`  applies to the column `b`, while the averages are taken over the values of the column `a`. These two columns have been indipendantly filled with data.

Mind that in the first statement, above, `anon_avg` differs ~10% from `avg` and `anon_avg_with_bounds`.
Therefore, I think that the bounds of `anon_avg` have not been computed correctly.

I came across the described behaviour after generating simple test data and playing with it.

The schema of the table is given by

```
create table A(ID integer, A numeric(4,2), B numeric(4,2), C numeric(4,2), D numeric(4,2), E numeric(4,2), F numeric(4,2));
```

It has been filled with uniformly distributed independant data being generated by the simple python script

```
import sys
from random import uniform as unif
import re

h=re.split(";|,", sys.argv[1])
n=int(sys.argv[2])
# print("ID;"+";".join(h))
print("\n".join([ str(i) + ";" + ";".join([str(round(unif(0, 10), 2)) for c in h]) for i in range(0, n) ]))
```

You may simply generate the test case by calling

`./gen_simple.sh A A,B,C,D,E,F 1000000`

using (after removing ".txt" and, maybe, adapting) the attached scripts.

Many regards and thank you for your wonderfull software :-)

Sebastian.

[gen_simple.sh.txt](https://github.com/google/differential-privacy/files/3761568/gen_simple.sh.txt)
[gen_simple_data.py.txt](https://github.com/google/differential-privacy/files/3761569/gen_simple_data.py.txt)
[gen_simple_import.py.txt](https://github.com/google/differential-privacy/files/3761570/gen_simple_import.py.txt)
[gen_simple_table.py.txt](https://github.com/google/differential-privacy/files/3761571/gen_simple_table.py.txt)

**Comment by dasmdasm** [comment]
Okay, so you're just taking an average numbers uniformly distributed between 0 and 10?

I think I see what's going on here. ANON_AVG uses automatic bounds determination logic, which tries to infer the maximum and minimum from the data. The implementation only picks multiples of two and doesn't perform well when you have very little data. My guess is that for the smaller dataset, you don't have enough values between 8 and 10 (looks like we'd expect there to be two) to get it to reliably pick an upper bound of 16. That means all your values will be truncated to a maximum of 8, and your average will come out smaller. When you add more data, it starts correctly detecting that there are values greater than 8 and ups the maximum to 16, meaning no data gets truncated (but twice as much noise is added).

The bottom line is that automatic bounds determination doesn't work all that well when you have a small number of data points, and that's expected behavior. As you note, providing more data fixes the issue.
